### PR TITLE
add densenet

### PR DESCRIPTION
**Pull request by qingzhouzhen**
Adapted it form [https://github.com/bruinxiong/densenet.mxnet/blob/master/symbol_densenet.py](https://github.com/bruinxiong/densenet.mxnet/blob/master/symbol_densenet.py)

### CODE REVIEW

**Comment by szha** [comment]
@qingzhouzhen thanks for the contribution. I also made a gluon version that can be found [here](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/model_zoo/vision/densenet.py). Since it was implemented using gluon's hybrid block, by feeding it with a symbol, you can get a symbol out already.

**Comment by qingzhouzhen** [comment]
do you suggest me to close this PR?

**Comment by szha** [comment]
The symbol-based image-classification is not connected with the gluon model zoo yet and it still has a lot of users, so there probably is value in keeping it. We will need to de-duplicate the codes and efforts soon.

**Comment by chinakook** [comment]
I think the old symbol is low-level and more compatible with Caffe and Tensorflow. However, Gluon API is high-level and more compatible with PyTorch and Keras. The former looks like C language and the latter looks like Java, so I suggest that keep both styles especially the former in the long run.

**Comment by szha** [comment]
For hybrid block they should be exactly the same. In that case the gluon interface is just a modular build tool.

**Comment by qingzhouzhen** [comment]
Ok,I will learn to use gluon, 
The training of densenet is really slow, the validation of top-1 is above 71% now(169 layers)
INFO:root:Epoch[42] Train-accuracy=0.796875
INFO:root:Epoch[42] Train-cross-entropy=0.746801
INFO:root:Epoch[42] Train-top_k_accuracy_5=0.955078
INFO:root:Epoch[42] Time cost=3112.315
INFO:root:Saved checkpoint to "densenet-models/densenet-2-0043.params"
INFO:root:Epoch[42] Validation-accuracy=0.710615
INFO:root:Epoch[42] Validation-cross-entropy=1.197285
INFO:root:Epoch[42] Validation-top_k_accuracy_5=0.902948

**Comment by piiswrong** [comment]
We can accept this for the sake of completeness. As along as you verify that this implementation is consistent with the gluon implementation

**Comment by qingzhouzhen** [comment]
The training of 169-layers is done, result as below:
INFO:root:Epoch[124] Batch [2450]       Speed: 107.01 samples/sec       accuracy=0.904687       cross-entropy=0.365293  top_k_accuracy_5=0.984219
INFO:root:Epoch[124] Batch [2300]       Speed: 106.73 samples/sec       accuracy=0.902656       cross-entropy=0.367414  top_k_accuracy_5=0.985313
INFO:root:Epoch[124] Batch [2100]       Speed: 106.68 samples/sec       accuracy=0.897656       cross-entropy=0.385241  top_k_accuracy_5=0.983594
INFO:root:Epoch[124] Batch [2500]       Speed: 106.51 samples/sec       accuracy=0.903438       cross-entropy=0.372258  top_k_accuracy_5=0.984219
INFO:root:Epoch[124] Train-accuracy=0.906250
INFO:root:Epoch[124] Train-cross-entropy=0.400677
INFO:root:Epoch[124] Train-top_k_accuracy_5=0.980469
INFO:root:Epoch[124] Time cost=3106.063
INFO:root:Saved checkpoint to "densenet-models/densenet-0125.params"
INFO:root:Epoch[124] Validation-accuracy=0.741744
INFO:root:Epoch[124] Validation-cross-entropy=1.160414
INFO:root:Epoch[124] Validation-top_k_accuracy_5=0.911830
I think the model is consistent with the gluon implementation, with little different such as BatchNorm layer, but it seems did not affect the result.

**Comment by piiswrong** [comment]
Why is there a difference? Which version is consistent with the paper?

**Comment by qingzhouzhen** [comment]
I think this one is more consistent with paper

**Comment by piiswrong** [comment]
@szha Is the gluon version different from paper?

**Comment by piiswrong** [comment]
This LGTM. If you would like to upload pretrained weight we can put it here: https://mxnet.incubator.apache.org/model_zoo/index.html?highlight=zoo

But first need to verify that performance is the same with original paper.

**Comment by qingzhouzhen** [comment]
Is it required to upload pretrained weight then to accept my PR?, because the pretrained weight is about 62M, which is restricted to upload in my company, and on the other hand, I had no PC with GPUs of my own. @piiswrong

File: example/image-classification/symbols/densenet.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    growth_rate : int
+        Number of output channels
+    """
+
+    if num_layers == 121:
+        units = [6, 12, 24, 16]
+    elif num_layers == 169:
+        units = [6, 12, 32, 32]
+    elif num_layers == 201:
+        units = [6, 12, 48, 16]
```
Comment (mikerossgithub) [inline example/image-classification/symbols/densenet.py]: To be consistent with https://arxiv.org/pdf/1608.06993.pdf you would want:
```
elif num_layers == 201:
  units = [6, 12, 48, 32]
```
I dont know if this is a typo here or in the paper.

**Comment by piiswrong** [comment]
@qingzhouzhen That's fine. Please address @mikerossgithub 's feedback and we can merge

**Comment by qingzhouzhen** [comment]
Thanks for finding my mistake @mikerossgithub , I have done it @piiswrong

**Comment by szha** [comment]
Sorry for the delayed response. @qingzhouzhen could you also add the config for 161, which has layers `[6, 12, 36, 24]`? The gluon model zoo already has it.

**Comment by qingzhouzhen** [comment]
OK, I have done it , would you check it?  I verified num_layer=161 can work by running some steps. @szha

**Comment by szha** [comment]
The master branch seems to be having some difficulty and @piiswrong is already managing the situation. Let's leave this to @piiswrong on when/where to merge this.

**Comment by szha** [comment]
@piiswrong shall we merge this?
